app/src/main/java/com/example/sharingang/ui/fragments/DetailedItemFragment.kt

```diff
@@ -198,80 +194,79 @@ class DetailedItemFragment : Fragment() {
-        itemViewModel.setRated(item)
-        itemViewModel.rated.observe(viewLifecycleOwner, {
-            val visibility =
-                if (!it && item.sold && currentUserProvider.getCurrentUserId() != null) View.VISIBLE
-                else View.GONE
+        itemViewModel.setReviews(item)
+        itemViewModel.reviews.observe(viewLifecycleOwner, {
+            val visibility = if (it.keys.contains(currentUserProvider.getCurrentUserId())
+                && currentUserProvider.getCurrentUserId() != null
+                && it[currentUserProvider.getCurrentUserId()!!]!!
+            ) View.VISIBLE
+            else View.GONE
             binding.ratingVisibility = visibility
         })
         binding.ratingButton.setOnClickListener {
-            val selectedOption: Int = binding.radioGroup.checkedRadioButtonId
-            if (selectedOption != -1) {
-                val rating = when (selectedOption) {
-                    binding.radioButton1.id -> 1
-                    binding.radioButton2.id -> 2
-                    binding.radioButton3.id -> 3
-                    binding.radioButton4.id -> 4
-                    binding.radioButton5.id -> 5
-                    else -> 0
-                }
-                viewModel.updateUserRating(item.userId, rating)
-                itemViewModel.rateItem(item)
+            val rating = when (binding.radioGroup.checkedRadioButtonId) {
+                binding.radioButton1.id -> 1
+                binding.radioButton2.id -> 2
+                binding.radioButton3.id -> 3
+                binding.radioButton4.id -> 4
+                binding.radioButton5.id -> 5
+                else -> 0
             }
+            viewModel.updateUserRating(item.userId, rating)
+            itemViewModel.updateReview(item, currentUserProvider.getCurrentUserId(), false)
         }
     }
 
-
-    private fun getButtonText(contains: Boolean): String {
-        return if (contains) getString(R.string.remove_wishlist)
-        else getString(R.string.add_wishlist)
-    }
-
     private fun updateWishlist() {
         viewModel.modifyWishList(args.item)
     }
 
-    /**
-     * Start the buying process
-     *
-     * On success, the item is marked as sold and the rating form is displayed
-     */
     private fun buyItem() {
-        binding.buyButton.isEnabled = false
-        lifecycleScope.launch {
-            val status = paymentProvider.requestPayment(args.item)
-            if (status) {
-                itemViewModel.sellItem(args.item)
-                binding.buyButton.visibility = View.GONE
-                initRating(args.item.copy(sold = true))
-            } else {
-                binding.buyButton.isEnabled = true
+        val quantity: Int = binding.quantity?.toIntOrNull() ?: 1
+        if (quantity > args.item.quantity) {
+            Toast.makeText(context, "Not enough available", Toast.LENGTH_SHORT).show()
+            binding.buyButton.isEnabled = true
+        } else {
```

Comment:
You should control that the quantity is more than 0.

Reply:
Check that it is more than 0. I know you said they can't put decimal or negative number, but I don't know how it works if they copy paste in the field. Maybe it doesn't allow to do so, but you should make sure they put more than 0, it wouldn't make sense to make a payment to buy 0 item.

Reply:
Fine I added the checks